### PR TITLE
Add Slack rate limiter

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

See https://github.com/dust-tt/decisions/issues/153.

This PR introduces a rate limiter for our Slackbot, restricting the number of unique Slack users that can interact with it within a 24-hour sliding window to the total of active members in a Dust workspace. This means in a workspace with two users (A and B), Slack users A and C could access the Slackbot. However, if B tries to access it afterwards, they’d be prompted to add more users to their Dust workspace. This limitation operates on a rolling 24-hour period.

Key updates in this PR include:
- Introducing a "generic" entity rate limiter.
  - Simple implementation that  uses a Redis sorted set where the entity id is stored alongside the current time as its score. For managing the 24-hour sliding window, we use [zrange](https://redis.io/commands/zrange/) to fetch all entity ids between two scores (lower bound being 24 hours ago and upper bound being the current moment). With each new entity action recorded, we also try to clean up any old keys that are no longer within the score range with [zremrangebyscore](https://redis.io/commands/zremrangebyscore/).
- Expose the count of active users through the front API. While okay-ish for an initial version, we might want to consider using a read replica for the front to manage such cases in the future. For now, we're caching these results for 15 minutes.
- Returning a slack message to upsell workspaces that have hit the rate limit.

image

## Risk

Pretty risky, tested locally with many users and mocking the maximum number of active users. Safe to rollback.

## Deploy Plan

Given the dependency on the Front api to get the count of active members within the workspace, we need to first deploy front and then deploy connectors.

### CODE REVIEW

File: connectors/src/connectors/slack/lib/workspace_rate_limit.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+) {
+  const slackClient = await getSlackClient(connector.id);
+  await slackClient.chat.postMessage({
+    channel: slackChannel,
+    blocks: [
+      {
+        type: "section",
+        text: {
+          type: "mrkdwn",
+          text: `Your Slack integration is limited to the number of active members in your workspace, which is currently ${maxAllowed} users. To allow more users to access the app, please increase your active member count.`,
```
Comment (spolu) [inline connectors/src/connectors/slack/lib/workspace_rate_limit.ts:80]: Deferring wording review to @gabhubert

Comment (spolu) [reply connectors/src/connectors/slack/lib/workspace_rate_limit.ts:80]: As an example "Increase your active member count" is not super clear? It also talks to the "admin" but the person there is the user.

Maybe it should be something more along the lines of "sign-in on Dust" ?

File: front/pages/api/v1/w/[wId]/members/count.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+): Promise<void> {
+  const keyRes = await getAPIKey(req);
+  if (keyRes.isErr()) {
+    return apiError(req, res, keyRes.error);
+  }
+  const { auth } = await Authenticator.fromKey(
+    keyRes.value,
+    req.query.wId as string
+  );
+
```
Comment (spolu) [inline front/pages/api/v1/w/[wId]/members/count.ts:24]: Maybe restrict to system API keys?

**Review by spolu — COMMENTED** [review]
Code LGTM \o/

But let's fine-tune wording with @gabhubert before shipping :+1:

File: connectors/src/connectors/slack/lib/workspace_rate_limit.ts (hunk 1 above)
Comment (flvndvd) [reply connectors/src/connectors/slack/lib/workspace_rate_limit.ts:80]: Oh yeah sign-in + auto-join domain = 😁.

File: front/pages/api/v1/w/[wId]/members/count.ts (hunk 2 above)
Comment (flvndvd) [reply front/pages/api/v1/w/[wId]/members/count.ts:24]: Fixed.

**Review by spolu — COMMENTED** [review]
@flvndvd we've meditated a bit with @gabhubert and after some discussion, we think the solution we picked is not the most optimal one:
- It's hard to explain
- It's annoying for actual members if they get rejected

We're considering one other very simple solution that we didn't cover in the original discussion (or maybe it was 2): Only respond to members of the workspace (matching on email) and provide a call to action otherwise:
- If auto-join is activated: a link
- Otherwise invite the user to "ask a Dust admin to add you to the workspace"

I believe you made an analysis of slack emails non matching members, do you confirm you used email as the matching key? Given the numbers it looks like a reasonable solution that will create FOMO for non members and will equally incentive joining the workspace + will not  hinder existing members. It's also simpler technically :)

Curious to hear your thoughts. Sorry for the back and forth on that. @gabhubert feel free to chime in if I missed or made up anything.

**Comment by flvndvd** [comment]
Closing in favor of https://github.com/dust-tt/dust/pull/3643.

**Pull request closed without merging**